numba/cpython/randomimpl.py

```diff
@@ -1586,3 +1586,60 @@ def dirichlet_impl(alpha, size=None):
         )
 
     return dirichlet_impl
+
+# ------------------------------------------------------------------------
+# np.random.noncentral_chisquare
+
+@overload(np.random.noncentral_chisquare)
+def noncentral_chisquare(df, nonc, size=None):
+
+    @register_jitable
+    def noncentral_chisquare_single(df, nonc):
+        # identical to numpy implementation 
```

Comment:
Am I correct in assuming that the NumPy implementation that this matches is:

https://github.com/numpy/numpy/blob/623bc1fae1d47df24e7f1e29321d0c0ba2771ce0/numpy/random/src/legacy/legacy-distributions.c#L119-L139

Reply:
If this is the case, it will be great if you can add an appropriate link to the comments.

Reply:
It's actually from:
https://github.com/numpy/numpy/blob/623bc1fae1d47df24e7f1e29321d0c0ba2771ce0/numpy/random/src/distributions/distributions.c
However, it seems that you're right that the default np.random calls legacy-distributions.c, so I'll update my implementation to reflect that.

Reply:
OK;  but if legacy is deprecated, maybe the better choice would be to move forwards? @stuartarchibald what do you think?

Reply:
IIRC it's legacy because of the new style RNG classes in NumPy, it still works fine. Suggest going for as close as possible to the NumPy impl as it makes debug/maintenance easier long term.

Reply:
After playing with it some more, I think the implementation from distributions.c is safer. For reference:

```c
double legacy_noncentral_chisquare(aug_bitgen_t *aug_state, double df,
                                   double nonc) {
  double out;
  if (nonc == 0) {
    return legacy_chisquare(aug_state, df);
  }
  if (1 < df) {
    const double Chi2 = legacy_chisquare(aug_state, df - 1);
    const double n = legacy_gauss(aug_state) + sqrt(nonc);
    return Chi2 + n * n;
  } else {
    const long i = random_poisson(aug_state->bit_generator, nonc / 2.0);
    out = legacy_chisquare(aug_state, df + 2 * i);
    /* Insert nan guard here to avoid changing the stream */
    if (npy_isnan(nonc)){
      return NPY_NAN;
    } else {
    return out;
    }
  }
}
```

Putting the np.nan return there doesn't play well with numba. Otherwise, however, the implementations are the same. Nonetheless, I added a comment referencing distributions.c.